azurelinuxagent/common/cgroupstelemetry.py

```diff
@@ -152,34 +142,38 @@ def cleanup():
 
 class CgroupMetrics(object):
     def __init__(self):
-        self._current_memory_usage = Metric()
-        self._max_memory_levels = Metric()
-        self._current_cpu_usage = Metric()
+        self._memory_usage = Metric()
+        self._max_memory_usage = Metric()
+        self._cpu_usage = Metric()
         self.marked_for_delete = False
 
-    def _add_memory_usage(self, metric):
-        self._current_memory_usage.append(metric[0].value)
-        self._max_memory_levels.append(metric[1].value)
-
-    def _add_cpu_usage(self, metric):
-        self._current_cpu_usage.append(metric[0].value)
-
-    def add_new_data(self, controller, metric):
-        if metric:
-            if controller == "cpu":
-                self._add_cpu_usage(metric)
-            elif controller == "memory":
-                self._add_memory_usage(metric)
+    def collect_data(self, cgroup):
+        # noinspection PyBroadException
+        try:
+            if cgroup.controller == "cpu":
+                self._cpu_usage.append(cgroup.get_cpu_usage())
+            elif cgroup.controller == "memory":
+                self._memory_usage.append(cgroup.get_memory_usage())
+                self._max_memory_usage.append(cgroup.get_max_memory_usage())
             else:
                 raise CGroupsException('CGroup controller {0} is not supported'.format(controller))
+        except Exception as e:
+            if not isinstance(e, (IOError, OSError)) or e.errno != errno.ENOENT:
+                logger.periodic_warn(logger.EVERY_HALF_HOUR, 'Could not collect metrics for cgroup {0}. Error : {1}'.format(cgroup.path, ustr(e)))
+
+    def get_memory_usage(self):
+        return self._memory_usage
+
+    def get_max_memory_usage(self):
+        return self._max_memory_usage
 
-    def get_metrics(self):
-        return self._current_memory_usage, self._max_memory_levels, self._current_cpu_usage
+    def get_cpu_usage(self):
```

Comment:
get_metrics was returning a list of values, replaced it with explicit methods for each value